Cython/Compiler/Code.py

```diff
@@ -2060,131 +2062,94 @@ def entry_as_pyobject(self, entry):
         type = entry.type
         if (not entry.is_self_arg and not entry.type.is_complete()
             or entry.type.is_extension_type):
-            return "(PyObject *)" + entry.cname
+            return "(PyObject *)" + entry.cname # FIXME need to implement this
         else:
             return entry.cname
 
     def as_pyobject(self, cname, type):
         from .PyrexTypes import py_object_type, typecast
         return typecast(py_object_type, type, cname)
 
-    def put_gotref(self, cname):
-        self.putln("__Pyx_GOTREF(%s);" % cname)
+    def put_gotref(self, cname, type):
+        self.putln(type.generate_gotref(cname))
```

Comment:
In all of these methods, the `putln()` call should be avoided if the type decides that there is nothing to write.

Reply:
I did add an `if None` right at the start of `putln` to try to account for this because otherwise all these functions ended up with duplicated boilerplate (which we're definitely trying to avoid in other place).

Maybe it'd be better to leave `putln` as it was, but add `putln_if_not_none` and then use that instead?

Reply:
There is a use case for `putln("")` to write a blank line, and I wouldn't rely on the result of `type.generate_*()` to always end up directly in `putln()`. So `putln()` (or a similar method) is the wrong place to handle this.

Why not pass the `code` writer on into the type method, and let that decide completely (by calling `putln()`) about how and what to generate? That has the advantage that the actual code generation is more encapsulated between code writer and type, so that it can be moved around between the two without caring about other usages elsewhere. Currently, anyone could call `type.generate_*()` and mishandle the result by accident.